File: plugin-sdk-ts/src/launch.ts
/**
 * Launch / environment activation module.
 *
 * Execute programs with specific environment configurations
 * and manage environment activation.
 *
 * Requires: `launch` permission (plus `env_read` for queries).
 */
import { callHostJson } from './host';
import type {
  ProcessResult,
  EnvActivationInfo,
  ActivationResult,
} from './types';

/** Execute a command with a specific environment activated. Requires: launch + process_exec */
export function withEnv(
  command: string,
  args: string[],
  envType: string,
  version?: string,
): ProcessResult {
  return callHostJson<ProcessResult>(
    'cognia_launch_with_env',
    JSON.stringify({ command, args, envType, version: version ?? null }),
  );
}

/** Get environment activation info. Requires: env_read */
export function getEnvInfo(envType: string): EnvActivationInfo {
  return callHostJson<EnvActivationInfo>(
    'cognia_launch_get_env_info',
    JSON.stringify({ envType }),
  );
}

/** Locate a program in PATH with optional env context. Requires: env_read */
export function whichProgram(command: string, envType?: string): string | null {
  return callHostJson<string | null>(
    'cognia_launch_which_program',
    JSON.stringify({ command, envType: envType ?? null }),
  );
}

/** Activate a specific environment version. Requires: launch */
export function activate(envType: string, version: string): ActivationResult {
  return callHostJson<ActivationResult>(
    'cognia_launch_activate',
    JSON.stringify({ envType, version }),
  );
}

File: plugin-sdk/src/launch.rs
//! Launch / environment activation module.
//!
//! Execute programs with specific environment configurations.

use crate::host;
use crate::types::*;
use extism_pdk::*;

/// Execute a command with a specific environment activated. Requires: launch + process_exec
pub fn with_env(command: &str, args: &[&str], env_type: &str, version: Option<&str>) -> Result<ProcessResult, Error> {
    let input = serde_json::json!({
        "command": command,
        "args": args,
        "envType": env_type,
        "version": version,
    }).to_string();
    let result = unsafe { host::cognia_launch_with_env(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get environment activation info. Requires: env_read
pub fn get_env_info(env_type: &str) -> Result<EnvActivationInfo, Error> {
    let input = serde_json::json!({ "envType": env_type }).to_string();
    let result = unsafe { host::cognia_launch_get_env_info(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Locate a program in PATH with optional env context. Requires: env_read
pub fn which_program(command: &str, env_type: Option<&str>) -> Result<Option<String>, Error> {
    let input = serde_json::json!({ "command": command, "envType": env_type }).to_string();
    let result = unsafe { host::cognia_launch_which_program(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Activate a specific environment version. Requires: launch
pub fn activate(env_type: &str, version: &str) -> Result<ActivationResult, Error> {
    let input = serde_json::json!({ "envType": env_type, "version": version }).to_string();
    let result = unsafe { host::cognia_launch_activate(input)? };
    Ok(serde_json::from_str(&result)?)
}
